tests: refactored fs watch_dir tests for stability

whitlockjc opened the pull request:
`fs_event_watch_dir` and `fs_event_watch_dir_recursive` could fail randomly
due to the way in which the tests were written.  Originally timers were
used to create, remove and recreate the test files but this could lead
to a race condition if the timeout used to delete the test files ran
before all file creation fs events were handled.  On top of that, the
file removal timer scheduled another timer to recreate the test files
and that timer's timeout could also lead to the same condition.

The refactoring removed timers for the removal/recreation of the test
files and instead the fs event callback was updated to have the
necessary logic to drive the test file removal.  We no longer recreate
the test files since it appears to be unnecessary.

With the current PR, I can run 1000 tests for `fs_event_watch_dir` and
`fs_event_watch_dir_recursive` without a single failure, where prior to this
PR they would fail 14/100 on average.

Fixes #30

whitlockjc commented:
/cc @saghul @indutny @misterdjules

whitlockjc commented:
One thing I forgot to mention, or maybe it wasn't clear, is that the original test appeared to do two cycles of creating files and then removing the files.  This current PR only does one cycle but updating it to do two should be trivial.  It's your call as to whether I add the extra cycle back or not.

saghul commented:
I'm afk until Monday. Will take a look then! Thanks!
On Aug 15, 2015 00:09, "Jeremy Whitlock" <notifications@github.com> wrote:

> /cc @saghul <https://github.com/saghul> @indutny
> <https://github.com/indutny> @misterdjules
> <https://github.com/misterdjules>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/libuv/libuv/pull/480#issuecomment-131252036>.
>

saghul commented on test/test-fs-event.c:
```diff
@@ -184,38 +144,76 @@ static const char* fs_event_get_filename_in_subdir(int i) {
+    fs_event_unlink_files(handle);
+  } else if (fs_event_cb_called == fs_event_created + fs_event_removed) {
+    /* Once we've processed all create and delete events, stop watching */
+    ASSERT(0 == uv_fs_event_stop(handle));
```
This is not really necessary, since calling `uv_close` will trigger a stop.

whitlockjc replied:
Makes sense.  I didn't write the original test so I didn't want to change it but I will remove it if you think it's unnecessary.

saghul commented:
CI: https://jenkins-iojs.nodesource.com/view/libuv/job/libuv+any-pr+multi/130/

saghul commented:
Not sure if it was failing before, but SmartOS and FreeBSD are failing:

SmartOS:
~~~~
`fs_event_watch_dir` failed: exit code 134
Output from process `fs_event_watch_dir`:
Assertion failed in test/test-fs-event.c on line 175: events == UV_RENAME
~~~~

FreeBSD:
~~~~
%  86|+ 216|-   7|T   0|S  12]: fs_event_watch_dir
`fs_event_watch_dir` failed: timeout
Output from process `fs_event_watch_dir`:
fs_event_cb_dir_multi_file called: 1
  created: 16
  removed: 0
~~~~

whitlockjc commented:
I will remove the unnecessary stuff and test on the other environments.  I will update you.

saghul commented:
Thanks! Ping me me if you need me to trigger the CI.

whitlockjc commented:
I have updated the PR.  I removed my debugging output *(sorry)*, removed the unnecessary assertions @saghul mentioned and updated the assertion in the `fs_event_cb_dir_multi_file` callback so we do not have assertion failures on SmartOS.  Fixing the assertion on SmartOS does not fix the tests completely as it will timeout like it does on FreeBSD but these tests were failing prior to my PR and appear to be a deeper libuv issue than a test race condition, since that is now gone.  *(A quick debugging session shows that on FreeBSD and SmartOS that the `fs_event_cb_dir_multi_file` callback is only invoked once instead instead of once per created/removed file.)*

saghul commented:
CI: https://jenkins-iojs.nodesource.com/view/libuv/job/libuv+any-pr+multi/132/

whitlockjc commented:
I can confirm I've somehow broken something.  I don't know how but it seems that way.  If I build `v1.x` on FreeBSD without my change, all 16 events are processed prior to the timeout whereas with my change, I only see 1 event processed prior to the timeout.  I'll get this sorted and will update the PR.

whitlockjc commented:
I've found something.  On FreeBSD and SmartOS, the tests were failing prior to this PR.  But after looking at the CI environment, it seems something changed for `fs_event_watch_dir` because it use to pass on CentOS 7 but now it's failing there.  I did notice on FreeBSD that prior to this PR, all create events were handled but that is not the case now.  That being said, I'll chase this down and update the PR.

whitlockjc commented:
I've cleaned up the sources a little to remove some compiler warnings on non-OSX/Windows.  Progress:

**FreeBSD**

FreeBSD was failing prior to this PR but I wanted to see if there was something I could do about it.  What I've noticed is that the fs event callback `fs_event_cb_dir_multi_file` is called only once even though the test creates 16 files, which should result in 16 separate `fs_event_cb_dir_multi_file` invocations.  `dtruss -e -d -f ./test/run-tests fs_event_watch_dir` shows only 3 `kevent` calls and that doesn't seem right but it seems to match up with what I'm seeing in my debugging.  On a whim, I removed the `EV_ONESHOT` from https://github.com/libuv/libuv/blob/v1.x/src/unix/kqueue.c#L97 and the test passes but I know that is not the fix, and it could very well be a red herring.  I'm working with @indutny on this but I don't know how much of his time I'll have.

**CentOS**

Building an environment as we speak.  No idea why it would stop passing due to these changes but I'll find out soon enough.

**SmartOS**

SmartOS was failing prior to this PR as well and it has similar symptoms to FreeBSD.  I will circle around to SmartOS when I make progress on one of the environments above.

whitlockjc commented:
I've merged in `v1.x` changes and resolved conflicts.  I'm testing on CentOS right now.

whitlockjc commented:
I just tested locally on CentOS 7 and `fs_event_watch_dir` passes without issue.  Testing FreeBSD again.

whitlockjc commented:
I think we're back in the state we were prior to this PR.  FreeBSD and SmartOS are failing as they were before and I think it might be a bigger issue.

@saghul Can you kick off a new CI build to make sure CentOS7 is passing again?

saghul commented:
Thanks again for the effort you're putting  on this. I'll trigger the CI as
soon as I'm back home.
On Aug 18, 2015 5:33 PM, "Jeremy Whitlock" <notifications@github.com> wrote:

> I think we're back in the state we were prior to this PR. FreeBSD and
> SmartOS are failing as they were before and I think it might be a bigger
> issue.
>
> @saghul <https://github.com/saghul> Can you kick off a new CI build to
> make sure CentOS7 is passing again?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/libuv/libuv/pull/480#issuecomment-132251855>.
>

whitlockjc commented:
No problem.  If you get a few minutes when you get home, ping me in IRC so we can chat.

indutny commented:
I have investigated the FreeBSD side. I'm afraid this is just how it works: it coalesces events into one `knote` and returns them at `kevent()` call.

I think we could do following thing - make it create and unlink files in different event-loop ticks. Does it make sense?

whitlockjc commented:
Yes.  I will update the PR when I've got that approach implemented.  Thanks for your help.

whitlockjc commented:
The latest version of the PR has `fs_event_watch_dir` passing on FreeBSD, which should allow us to check that one off of #128 if/when this gets merged, and it is also passing on SmartOS now.  *(Both of these were failing prior to this PR.)*

Once @indutny verified that the fs event callback `fs_event_watch_dir` was only being called once and was by design *(expected)*, we realized the only way for this test to pass as written was to create one file per event loop tick and remove one file per event loop tick to make sure that the proper number of fs events were emitted for the created and removed files.

As it stands, this PR not only makes the `fs_event_watch_dir` and `fs_event_watch_dir_recursive` stable, it also fixes `fs_event_watch_dir` on FreeBSD, SmartOS and possibly others where it might had been failing.  Thanks a lot @indutny and @misterdjules.

saghul commented:
@whitlockjc sorry, yesterday ended up not working for me. CI: https://jenkins-iojs.nodesource.com/view/libuv/job/libuv+any-pr+multi/133/

saghul commented:
@whitlockjc can you please rebase again? I pushed a fix for the broken build on Windows.

whitlockjc commented:
Done.

saghul commented:
CI: https://jenkins-iojs.nodesource.com/view/libuv/job/libuv+any-pr+multi/135/

saghul commented:
Great job @whitlockjc! LGTM :+1:

whitlockjc commented:
Thanks to all that helped: @indutny and @misterdjules  Two excellent mentors.

saghul commented:
Landed in :sparkles: 8326470 :sparkles: Much obliged, @whitlockjc!

whitlockjc commented:
WOOHOO!
